test/smoother/test_smoother.cpp

```diff
@@ -0,0 +1,329 @@
+    int output_seq_len, int seq_stride, std::vector<float> &mu_x,
+    std::vector<float> &sigma_x)
+/* Get dataloader for input and output data.
+
+Args:
+
+Returns:
+    dataset: Dataloader
+ */
+{
+    Dataloader db;
+    int num_outputs = output_col.size();
+    std::vector<float> x(num_features * num_data, 0), cat_x;
+
+    // Load input data from csv file that contains the input & output data.
+    // NOTE: csv file need to have a header for each columns
+    for (int i = 0; i < data_file.size(); i++) {
+        read_csv(data_file[i], x, num_features, true);
+        cat_x.insert(cat_x.end(), x.begin(), x.end());
+    };
+
+    // Compute sample mean and std for dataset
+    if (mu_x.size() == 0) {
+        mu_x.resize(num_features, 0);
+        sigma_x.resize(num_features, 1);
+        compute_mean_std(cat_x, mu_x, sigma_x, num_features);
+    }
+    std::vector<float> mu_y(num_outputs, 0);
+    std::vector<float> sigma_y(num_outputs, 1);
+    if (data_norm) {
+        normalize_data(cat_x, mu_x, sigma_x, num_features);
+        for (int i = 0; i < num_outputs; i++) {
+            mu_y[i] = mu_x[output_col[i]];
+            sigma_y[i] = sigma_x[output_col[i]];
+        }
+    }
+
+    // Create rolling windows
+    int num_samples =
+        (cat_x.size() / num_features - input_seq_len - output_seq_len) /
+            seq_stride +
+        1;
+    std::vector<float> input_data(input_seq_len * num_features * num_samples);
+    std::vector<float> output_data(output_seq_len * num_outputs * num_samples);
+    create_rolling_windows(cat_x, output_col, input_seq_len, output_seq_len,
+                           num_features, seq_stride, input_data, output_data);
+
+    // Set data to output variable
+    db.x = input_data;
+    db.mu_x = mu_x;
+    db.sigma_x = sigma_x;
+    db.nx = num_features * input_seq_len;
+
+    db.y = output_data;
+    db.mu_y = mu_y;
+    db.sigma_y = sigma_y;
+    db.ny = num_outputs * output_seq_len;
+    db.num_data = num_samples;
+
+    return db;
+}
+
+void replace_with_prediction(std::vector<float> &mu_sequence,
+                             std::vector<float> &x) {
+    // Replace nan in input x by the lstm_prediction
+    for (size_t i = 0; i < x.size(); ++i) {
+        if (x[i] == 0) {
+            x[i] = mu_sequence[i];
+        }
+    }
+}
+
+void add_prediction_to_sequence(size_t input_seq_len, float m_pred,
+                                std::vector<float> &mu_sequence) {
+    // Add new prediction to mu_sequence
+    mu_sequence.push_back(m_pred);
+
+    // Ensure mu_sequence has only the last `input_seq_len` elements
+    if (mu_sequence.size() > input_seq_len) {
+        mu_sequence.erase(mu_sequence.begin(),
+                          mu_sequence.end() - input_seq_len);
+    }
+}
+
+void smoother_v1()
+/**/
+{
+    // Data
+    int num_train_data = 240;
+    int num_test_data = 48;
+    std::vector<int> output_col{0};
+    int num_features = 1;
+    std::vector<std::string> x_train_path{
```

Comment:
@van-dai-vuong this file does not exist. I've changed to "x_train_sin_smoother_zero.csv" but it has the following error
```shell
Program received signal SIGSEGV, Segmentation fault.
0x0000555555627c48 in smooth_cell_states (num_timestep=216, num_states=40, cov_cc=std::vector of length 8640, capacity 8640 = {...}, mu_c_priors=std::vector of length 8640, capacity 8640 = {...}, var_c_priors=std::vector of length 8640, capacity 8640 = {...}, mu_c_posts=std::vector of length 8640, capacity 8640 = {...}, var_c_posts=std::vector of length 8640, capacity 8640 = {...}, mu_c_smooths=std::vector of length 8640, capacity 8640 = {...}, var_c_smooths=std::vector of length 8640, capacity 8640 = {...}) at /home/ha/Tagi/cuTAGI/src/slstm_layer.cpp:138
138                 float tmp = cov_cc[next] / var_c_priors[next];
```
Could you please make sure the C++ test properly? I only rely on this code to check the bugs e.g., memory leak

Reply:
I have added the *csv file. Now the c++ test runs properly. Sorry.